client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoPreparedStatement.java

```diff
@@ -109,14 +109,29 @@ public class TrinoPreparedStatement
     private final String statementName;
     private final String originalSql;
     private boolean isBatch;
+    private boolean prepareStatementExecuted;
 
     TrinoPreparedStatement(TrinoConnection connection, Consumer<TrinoStatement> onClose, String statementName, String sql)
             throws SQLException
     {
         super(connection, onClose);
         this.statementName = requireNonNull(statementName, "statementName is null");
         this.originalSql = requireNonNull(sql, "sql is null");
-        super.execute(format("PREPARE %s FROM %s", statementName, sql));
+        if (connection().isUseLegacyPreparedStatements()) {
+            super.execute(format("PREPARE %s FROM %s", statementName, sql));
+            prepareStatementExecuted = true;
+        }
+    }
+
+    private void prepareStatementIfNecessary()
+            throws SQLException
+    {
+        if (prepareStatementExecuted) {
+            return;
+        }
+
+        super.execute(format("PREPARE %s FROM %s", statementName, originalSql));
```

Comment:
we seem to execute this unconditionally? Shouldn't this only be done if legacy prepared statements are requested?

Or is this because some methods on `TrinoPreparedStatement` don't work with `EXECUTE IMMEDIATE`? (e.g. `getMetaData`, `getParameterMetadata`)?

This for sure deserves a comment since it's not obvious.

Reply:
This line should be executed only when 1) `isUseLegacyPreparedStatements` is disabled and 2) `getMetadata()` or `getParameterMetadata()` are called. But the location of the function is probably leading to confusion.

Reply:
With the exception of `testGetClientTypeSignatureFromTypeString`, all the tests in `TestJdbcPreparedStatement` were modified to use both legacy and new prepared statements (and it indeed confirmed that the two metadata methods were failing). Can you think of another type of test that should be added?

Reply:
so it seems it's covered by existing tests. Thanks for explaining. no change requested here.